src/search-manager/hooks.ts

```diff
@@ -0,0 +1,133 @@
+import React from 'react';
+import {
+  type FromStringFn,
+  type ToStringFn,
+  useStateWithUrlSearchParam,
+} from '../hooks';
+
+/**
+ * Typed hook that returns useState if skipUrlUpdate,
+ * or useStateWithUrlSearchParam otherwise.
+ *
+ * Function is overloaded to accept simple Type or Type[] values.
+ *
+ * Provided here to reduce some code overhead in SearchManager.
+ */
+export function useStateOrUrlSearchParam<Type>(
+  defaultValue: Type[],
+  paramName: string,
+  fromString: FromStringFn<Type>,
+  toString: ToStringFn<Type>,
+  skipUrlUpdate?: boolean,
+): [value: Type[], setter: React.Dispatch<React.SetStateAction<Type[]>>];
+export function useStateOrUrlSearchParam<Type>(
+  defaultValue: Type,
+  paramName: string,
+  fromString: FromStringFn<Type>,
+  toString: ToStringFn<Type>,
+  skipUrlUpdate?: boolean,
+): [value: Type, setter: React.Dispatch<React.SetStateAction<Type>>];
+export function useStateOrUrlSearchParam<Type>(
+  defaultValue: Type | Type[],
+  paramName: string,
+  fromString: FromStringFn<Type>,
+  toString: ToStringFn<Type>,
+  skipUrlUpdate?: boolean,
+): [value: Type | Type[], setter: React.Dispatch<React.SetStateAction<Type | Type[]>>] {
+  const useStateManager = React.useState<typeof defaultValue>(defaultValue);
+  const urlStateManager = useStateWithUrlSearchParam<typeof defaultValue>(
+    defaultValue,
+    paramName,
+    fromString,
+    toString,
+  );
+
+  return skipUrlUpdate ? useStateManager : urlStateManager;
+}
+
+/**
+ * Helper class for managing Block + Problem type states.
+ *
+ * We use a class to store both Block and Problem types together because
+ * their behaviour is tightly intertwined: e.g if Block type "problem" is
+ * selected, that means all available Problem types are also selected.
+ *
+ */
+export class TypesFilterData {
+  #blocks = new Set<string>();
+
+  #problems = new Set<string>();
+
+  static #sanitizeType = (value: string | null | undefined): string | undefined => (
+    (value && /^[a-z0-9._-]+$/.test(value))
+      ? value
+      : undefined
+  );
+
+  static #sep1 = ','; // separates the individual types
+
+  static #sep2 = '|'; // separates the block types from the problem types
```

Comment:
@pomegranited NIT: we can replace these separators with browsers ability to parse multiple query params with same name. But I don't want to block this as we can update it in a separate PR in future (if we want to). Currently, I cannot find any real problems with this approach and it works perfectly.

Reply:
Hi @navinkarkera -- I'd love to do this, but I'm not sure how, since there's two different "types" that we're tracking here -- block and problem types -- I can't just use a simple `?type=html&type=checkboxes&...` solution. I had to merge them into a single object because they're so entwined (see [`FilterByBlockType`](https://github.com/open-craft/frontend-app-authoring/blob/jill/fal-3984-sharable-urls/src/search-manager/FilterByBlockType.tsx)), and so when I then tried to manage their state with URLSearchParams, I got all kinds of terrible react errors.

I'm open to ideas though -- I just can't see a clear way to doing this.

Reply:
@pomegranited I was thinking about using a different param name for problem types for example: 
`?type=html&type=checkboxes&problemType=multiplechoiceresponse&problemType=optionresponse`
But if it is causing issues with react and requires a lot of time to implement, we can skip it for now.

Reply:
It won't take much time to split these params -- it's just a reversion back to where they were before. And it's possible the stuff I did to workaround the URLSearchParams state issue has fixed the react problems I was having when I combined them.

@ChrisChV what do you think?

Reply:
I think Jill's solution is fine for now. I'm not one to risk errors in reverting when they've already been fixed. In any other case, it could be done later. What do you think?